### PR TITLE
CODENVY-505 Rework permissions mechanism for stacks, recipes and system perms.

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
Add JPA-based storage for Stack, Recipe, System permissions and Workers.

### What issues does this PR fix or reference?
CODENVY-505

### Tests written?
Yes

### Docs requirements?
None.

### CODE REVIEW

File: wsmaster/codenvy-hosted-api-permission-shared/pom.xml [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                        <artifactId>codenvy-hosted-api-permission-shared</artifactId>
+                        <version>${project.version}</version>
+                    </dependency>
+                </dependencies>
+                <configuration>
+                    <dtoPackages>
+                        <package>com.codenvy.api.permission.shared.dto</package>
+                    </dtoPackages>
+                    <outputDirectory>${dto-generator-out-directory}</outputDirectory>
+                    <genClassName>com.codenvy.api.permission.server.dto.DtoServerImpls</genClassName>
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-permission-shared/pom.xml]: We should not generate server implementation for dto in shared module. Please move it to server code module

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/PermissionsManager.java [hunk 2]
```diff
@@ -75,117 +76,118 @@ public PermissionManager(Set<PermissionsStorage> storages) throws ServerExceptio
...
 
     /**
-     * @param user
-     *         user id
-     * @param domain
-     *         domain id
-     * @param instance
-     *         instance id
+     * @param userId
+     *         userId id
```
Comment (sleshchenko) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/PermissionsManager.java]: `userId id`

File: wsmaster/codenvy-hosted-api-permission-shared/pom.xml (hunk 1 above)
Comment (mshaposhnik) [inline wsmaster/codenvy-hosted-api-permission-shared/pom.xml]: Ok

File: wsmaster/codenvy-hosted-api-permission-shared/src/main/java/com/codenvy/api/permission/shared/model/Permissions.java [hunk 3]
```diff
@@ -29,20 +29,20 @@
      *
      * <p>Note: also supported '*' for marking all users
```
Comment (mshaposhnik) [inline wsmaster/codenvy-hosted-api-permission-shared/src/main/java/com/codenvy/api/permission/shared/model/Permissions.java]: I think this Note isn't actual anymore?

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/model/impl/SystemPermissionsImpl.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+/**
+ * @author Max Shaposhnik
+ */
+@Entity(name = "SystemPermissions")
+@NamedQueries(
+        {
+                @NamedQuery(name = "SystemPermissions.getByUserId",
+                            query = "SELECT recipe " +
+                                    "FROM RecipePermissions recipe " +
+                                    "WHERE recipe.userId = :userId ")
```
Comment (mshaposhnik) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/model/impl/SystemPermissionsImpl.java]: LoL

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/JpaSystemPermissionsDao.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        if (existed.isEmpty()) {
+            throw new NotFoundException(format("System permissions for user '%s' not found", userId));
+        }
+        return existed.get(0);
+    }
+
+    @Override
+    @Transactional
+    public List<SystemPermissionsImpl> getByInstance(String instanceId) throws ServerException {
+        throw new ServerException("This operation is not supported for system permissions.");
```
Comment (mshaposhnik) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/JpaSystemPermissionsDao.java]: Not very appropriate exc. Maybe just return null?

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/SystemDomain.java [hunk 6]
```diff
@@ -44,4 +47,9 @@ public SystemDomain(@Named(SYSTEM_DOMAIN_ACTIONS) Set<String> allowedActions) {
                     .collect(Collectors.toList()),
               false);
     }
+
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/SystemDomain.java]: javadoc

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+import javax.persistence.EntityManager;
+import java.io.IOException;
+import java.util.List;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * @author Max Shaposhnik
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java]: javadoc

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/model/impl/SystemPermissionsImpl.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+import com.codenvy.api.permission.server.SystemDomain;
+
+import javax.persistence.Entity;
+import javax.persistence.NamedQueries;
+import javax.persistence.NamedQuery;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * @author Max Shaposhnik
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/model/impl/SystemPermissionsImpl.java]: javadoc

File: wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/recipe/RecipePermissionsImpl.java [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+import javax.persistence.ManyToOne;
+import javax.persistence.NamedQueries;
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
+import javax.persistence.UniqueConstraint;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ *
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/recipe/RecipePermissionsImpl.java]: javadoc

File: wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/stack/StackPermissionsImpl.java [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+import javax.persistence.ManyToOne;
+import javax.persistence.NamedQueries;
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
+import javax.persistence.UniqueConstraint;
+import java.util.List;
+import java.util.Objects;
+
+
+/**
```
Comment (skabashnyuk) [inline wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/stack/StackPermissionsImpl.java]: javadoc

**Comment by skabashnyuk** [comment]
Add tck to codenvy-integration-tck-jpa

File: wsmaster/codenvy-hosted-api-permission/pom.xml [hunk 11]
```diff
@@ -67,23 +68,67 @@
...
+        </dependency>
+        <dependency>
+            <groupId>com.google.inject.extensions</groupId>
+            <artifactId>guice-persist</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.che.core</groupId>
+            <artifactId>che-core-api-machine</artifactId>
+            <scope>provided</scope>
```
Comment (akorneta) [inline wsmaster/codenvy-hosted-api-permission/pom.xml]: this dependency shouldn't be in `provided` scope

File: wsmaster/codenvy-hosted-api-permission/pom.xml [hunk 12]
```diff
@@ -67,23 +68,67 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.eclipse.che.core</groupId>
+            <artifactId>che-core-api-user</artifactId>
+            <scope>provided</scope>
```
Comment (akorneta) [inline wsmaster/codenvy-hosted-api-permission/pom.xml]: same with [this](https://github.com/codenvy/codenvy/pull/603/files#r76378438)

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+        requireNonNull(instanceId, "Instance identifier required");
+        requireNonNull(userId, "User identifier required");
+        requireNonNull(action, "Action name required");
+        T permissions;
+        try {
+            permissions = get(userId, instanceId);
+        } catch (NotFoundException e) {
+            return false;
+        }
+        return  permissions != null && permissions.getActions().contains(action);
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java]: as far as i understand there is no way to get null permissions

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/AbstractPermissionsDomain.java [hunk 14]
```diff
@@ -66,6 +66,18 @@ public Boolean isInstanceRequired() {
         return requiresInstance;
     }
 
+    /*
+    * Creates new instance of the entity related to this domain.
+    */
+    public T newInstance(String userId, String instanceId, List<String> allowedActions) throws ConflictException{
```
Comment (akorneta) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/AbstractPermissionsDomain.java]: fix formatting

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/JpaSystemPermissionsDao.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * JPA based implementation of system permissions DAO.
+ *
+ * @author Max Shaposhnik
+ */
+public class JpaSystemPermissionsDao extends AbstractPermissionsDao<SystemPermissionsImpl> {
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/JpaSystemPermissionsDao.java]: Does it have to be a singleton ?

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/JpaSystemPermissionsDao.java [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+    @Inject
+    public JpaSystemPermissionsDao(@Named(SystemDomain.SYSTEM_DOMAIN_ACTIONS) Set<String> allowedActions) throws IOException {
+        super(new SystemDomain(allowedActions));
+    }
+
+    @Override
+    @Transactional
+    public SystemPermissionsImpl get(String userId, String instanceId) throws ServerException, NotFoundException {
+        List<SystemPermissionsImpl> existed = getByUser(userId);
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/JpaSystemPermissionsDao.java]: why `existed` do you mean existing or existent ?

File: wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/recipe/RecipePermissionsImpl.java [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        if (!(obj instanceof RecipePermissionsImpl)) return false;
+        final RecipePermissionsImpl other = (RecipePermissionsImpl)obj;
+        return Objects.equals(userId, other.userId) &&
+               Objects.equals(recipeId, other.recipeId) &&
+               actions.equals(other.actions);
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/recipe/RecipePermissionsImpl.java]: `new RecipePermissionsImpl().equals(new RecipePermissionsImpl())` will produce npe at this point

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java [hunk 18]
```diff
@@ -0,0 +1,157 @@
...
+ */
+public abstract class AbstractPermissionsDao<T extends AbstractPermissions> implements PermissionsDao<T> {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractPermissionsDao.class);
+    private final AbstractPermissionsDomain<T> supportedDomain;
+
+
+    @Inject
+    protected Provider<EntityManager> managerProvider;
+
+    @Inject
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java]: Are you sure you need `@Inject` here?

File: wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+import java.util.List;
+import java.util.Set;
+
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * @author Max Shaposhnik
+ */
+public abstract class AbstractPermissionsDao<T extends AbstractPermissions> implements PermissionsDao<T> {
```
Comment (evoevodin) [inline wsmaster/codenvy-hosted-api-permission/src/main/java/com/codenvy/api/permission/server/jpa/AbstractPermissionsDao.java]: It uses JPA, so it looks like `Jpa` should be included to the class name

**Comment by evoevodin** [comment]
Did you consider an approach which simplifies permissions dao implementations ?
I see a few ways of doing that:
- use the same named queries for different entities 
e.g. `Recipe.getByInstanceAndUser` `Stack.getByInstanceAndUser` by specifying an entity class you can easily reuse the custom named queries for different implementations
- specify join columns for the default query, which will join an entity to the permissions and get all the needed information.

in this way dao implementations look really simple.
Can you please commend on that?

File: wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/jpa/JpaWorkerDao.java [hunk 20]
```diff
@@ -14,89 +14,98 @@
...
 /**
- * Jpa implementation of {@link WorkerDao},
+ * JPA based implementation of worker DAO.
  *
  * @author Max Shaposhnik
  */
 @Singleton
-public class JpaWorkerDao implements WorkerDao {
+public class JpaWorkerDao extends AbstractPermissionsDao<WorkerImpl> implements WorkerDao
+{
```
Comment (akorneta) [inline wsmaster/codenvy-hosted-api-workspace/src/main/java/com/codenvy/api/workspace/server/jpa/JpaWorkerDao.java]: fix formatting

File: wsmaster/codenvy-hosted-api-workspace/src/test/java/com/codenvy/api/workspace/server/spi/tck/RecipePermissionsDaoTest.java [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+        final Permissions result = dao.get(oldPermissions.getUserId(), oldPermissions.getInstanceId());
+
+        assertEquals(newPermissions, result);
+    }
+
+    @Test
+    public void shouldReturnsSupportedDomainsIds() {
+        assertEquals(dao.getDomain(), new TestDomain());
+    }
+
```
Comment (akorneta) [inline wsmaster/codenvy-hosted-api-workspace/src/test/java/com/codenvy/api/workspace/server/spi/tck/RecipePermissionsDaoTest.java]: remove extra lines

File: wsmaster/codenvy-hosted-api-workspace/src/test/java/com/codenvy/api/workspace/server/spi/tck/StackPermissionsDaoTest.java [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+        final Permissions result = dao.get(oldPermissions.getUserId(), oldPermissions.getInstanceId());
+
+        assertEquals(newPermissions, result);
+    }
+
+    @Test
+    public void shouldReturnsSupportedDomainsIds() {
+        assertEquals(dao.getDomain(), new TestDomain());
+    }
+
```
Comment (akorneta) [inline wsmaster/codenvy-hosted-api-workspace/src/test/java/com/codenvy/api/workspace/server/spi/tck/StackPermissionsDaoTest.java]: same with [this](https://github.com/codenvy/codenvy/pull/603/files#r76387003)

**Comment by mshaposhnik** [comment]
@evoevodin  We tried to re-use same search queries for different permission types, but this approach requires: 
- Setting instance class during abstract dao construction;
- Passing named query or it's content;
- Passing search parameter name and value (instaceId field name is different for each type, moreover, system permissions doesn't have it at all)
- Instances like Worker is potentially extendable, and queries may became more complex (including more fields etc)

So for this moment we decided to stay at overriding each type select methods using own queries.

**Comment by skabashnyuk** [comment]
ok

**Comment by evoevodin** [comment]
LGTM

**Pull request merged**